lib/components/user/user-account-screen.js

```diff
@@ -72,21 +66,93 @@ function cloneWithHomeAndWorkAsTopLocations (loggedInUser) {
+      // Reminder: When a new user progresses through the account steps,
+      // isNewUser(loggedInUser) will change to false as the database gets updated.
+      isNewUser: isNewUser(props.loggedInUser),
+
+      // Last number and last time we requested a code for (to avoid repeat SMS and not waste SMS quota).
+      lastPhoneNumberRequested: null,
+      lastPhoneRequestTime: null
+    }
+  }
+
+  _updateUserPrefs = async (userData, silentOnSucceed = false) => {
     // TODO: Change state of Save button while the update action takes place.
 
     // In userData.savedLocations, filter out entries with blank addresses.
     const newUserData = clone(userData)
     newUserData.savedLocations = newUserData.savedLocations.filter(({ address }) => address && address.length)
-    await this.props.createOrUpdateUser(newUserData)
+    await this.props.createOrUpdateUser(newUserData, silentOnSucceed)
 
     // TODO: Handle UI feedback (currently an alert() dialog inside createOrUpdateUser).
   }
 
+  /**
+   * Silently persists the user data upon accepting terms.
+   * Creating the user record before the user finishes the account creation steps
+   * is required by the middleware in order to perform phone verification.
+   *
+   * @param {*} userData The user data state to persist.
+   * @returns The new user id the the caller can use.
+   */
+  _handleCreateNewUser = async userData => {
+    await this._updateUserPrefs(userData, true)
+
+    return this.props.loggedInUser.id
+  }
+
   _handleExit = () => {
     // On exit, route to default search route.
     this.props.routeTo('/')
   }
+
+  /**
+   * Requests a phone verification SMS to the specified phone number.
+   * This handler is called when the user clicks "Verify my phone" after entering a new number.
+   * @returns The effective phone number and verification status.
+   */
+  _handleRequestPhoneVerificationSms = async numberToVerify => {
+    const { lastPhoneNumberRequested, lastPhoneRequestTime } = this.state
+    const timestamp = new Date()
+
+    // Request a new verification code if we are requesting a different number.
+    // or enough time has ellapsed since the last request (1 minute?).
+    // TODO: Should throttling be handled in the middleware?
+    if (lastPhoneNumberRequested !== numberToVerify ||
+        (lastPhoneRequestTime && lastPhoneRequestTime <= timestamp + 60000)) {
+      this.setState({
+        lastPhoneNumberRequested: numberToVerify,
+        lastPhoneRequestTime: timestamp
+      })
+      await this.props.requestPhoneVerificationSms(numberToVerify)
+    }
+
+    // Return the phone number and state from state.user.loggedInUser.
+    // If the SMS request was successful, the pending phone number and unverified status
+    // will be reflected in the state, otherwise they will be unchanged.
+    const { isPhoneNumberVerified, phoneNumber } = this.props.loggedInUser
```

Comment:
This makes me a little uneasy because this method relies on the `this.props` variable changing during the `await` on line 132. This seems to be way too much application logic being put into a component. Ideally, components should do as little application logic as possible and focus solely on rendering and delegating event handling to actions. Furthermore, the `lastPhoneNumberRequested` and `lastPhoneRequestTime` state variables are only used here and not passed along to any components to be rendered. To me this is another indication of variables that belong in the redux store (application state) instead of component state since they have no effect on how a component renders.

Reply:
I agree on moving `lastPhoneNumberRequested/Time` to the redux state. What would be a better way to capture the outcome of the actions? `loggedInUser` is already part of the redux state.

Reply:
last phone number states move to redux state in https://github.com/opentripplanner/otp-react-redux/pull/224/commits/bc43d29931e3907bca16067f26d718075ad0e6e3.